### ISSUE TITLE
Interaction between lists and paragraphs

### ISSUE THREAD

**Issue by Enivex** [typst/typst#3206]
Currently, a tight list attaches to the preceding paragraph when there is no whitespace, and a new paragraph is always created after the list.

![image](https://github.com/typst/typst/assets/2191754/35c5d8f8-5d9d-477c-a808-6383ce15ed5e)

In this case, this behavior is not desirable, because everything here is really part of the _same_ paragraph.

For comparison, the behavior in LaTeX is that the next line is also part of the paragraph (new paragragraphs are indented)
![image](https://github.com/typst/typst/assets/2191754/0c680a74-a9c4-4752-be7e-645d99a8c35a)

_unless_ whitespace is added
![image](https://github.com/typst/typst/assets/2191754/7f4900b4-5d8c-4e6d-8c17-1e6c92f32969)

**Comment by BrandonHenke** [comment]
This is related to how typst handles paragraphs. A list is a different element than a paragraph, so, no matter how many lines are placed between the list and the next piece of content, it will treat the next piece of content as its own paragraph.

A paragraph is only indented if it follows another paragraph (and has first-line-indent != 0 and some alignment stuff). Typst doesn't consider a list a paragraph. Hence, the content that comes after the list will be its own paragraph, and that paragraph will not be indented.

I've forked the repo, because I think the paragraph element is too limited as it is. I'm attempting to change that.

**Comment by Enivex** [comment]
I modified the title to reflect that this is a wider issue

**Comment by Enivex** [comment]
Another example: Figures that are not `top` or `bottom`

**Comment by laurmaedje** [comment]
I'm not really convinced that a figure really makes sense as part of a paragraph.

**Comment by BrandonHenke** [comment]
> I'm not really convinced that a figure really makes sense as part of a paragraph.

How sensical you or anyone else thinks it is doesn't matter.
This is a typesetting software.
Disallowing some sort of formatting because it doesn't make sense to someone would be limiting.

A figure is a block of content.
A paragraph is a string of connected ideas.
If one of those ideas is best given as an equation block, code block, figure block, etc., then discriminating between content blocks is limiting to the string of connected ideas used to form the paragraph.

**Comment by BrandonHenke** [comment]
I'm sorry if I seemed rude in my previous comment. I in no way meant it antagonistically. My apologies!

**Comment by OrangeX4** [comment]
For some people who need a workaround:

```typst
#set page(width: 30em)
#set par(first-line-indent: (amount: 2em, all: true))
#set math.equation(numbering: "(1.1)")
#set heading(numbering: "1.")

#lorem(10), for example,
#box($ integral x + y = z $)
shows that the integral of $x + y$ is $z$.

#lorem(20)
```

![Image](https://github.com/user-attachments/assets/dc13e0f2-3d7f-45f8-882b-24d7f37fb0e2)

**Comment by jakubkaczor** [comment]
@OrangeX4, is your “work-around” even necessary? I appear to get the same results without it, apart from the lack of indent at the beginning of the column.

**Comment by jakubkaczor** [comment]
The original post is about creating a new paragraph _after_ a block element, but I have the opposite problem with exposed formulas. Regardless of empty lines, I cannot get the first line indent for the text after a formula. I appear to get the same results without it. Here is an example of the issue.

```typst
#set page(paper: "a5", fill: white)
#set par(first-line-indent: 4em)

= Good

A function $f$ defined by the following formula.
$
  f(x) = c,
$
where $c in RR$.


= Bad

A function $f$ defined by the following formula.
$
  f(x) = c.
$

This is expected to be a new paragraph due to line break, but there is no first line indent.

$
  "This is expected to be a new paragraph. Is it?"
$
#lorem(15)
```
Results:

![Image](https://github.com/user-attachments/assets/d6c5a7c0-318f-4bb3-94db-c18e835d843b)

**Comment by Enivex** [comment]
Yes, that is probably more or less how "block" equations should've worked in the first place.

**Comment by Yangchengran** [comment]
I tried to use the #box function, but still need to par break when use set par justify to be true.

```typ
#set par(justify: true, first-line-indent: (amount:2em, all:true))
#set math.equation(numbering: "(1.1)")

This is a line
#box($ x_1 + x_2 = x_3 $)
This is a new line

#line(length:100%)

This is a line
\ #box($ x_1 + x_2 = x_3 $)
This is a new line
```

![Image](https://github.com/user-attachments/assets/712db6e5-a87c-472e-bd5e-c5b9500c7a2b)

**Comment by YoniKF** [comment]
Another problem with the `#box` solution is that it does not work when numbering is disabled, as the box becomes inline, for example,
```typst
#set page(width: 30em)
#set par(first-line-indent: (amount: 2em, all: true))
#set math.equation(numbering: none)

#lorem(10), for example,
#box[$ integral x + y = z $]
shows that the integral of $x + y$ is $z$.

#lorem(20)
```
gives
<img width="908" alt="Image" src="https://github.com/user-attachments/assets/ddd43e51-a59e-430a-ab92-8e069244701f" />
Any workaround for that?

**Comment by iwkjosec** [comment]
> Any workaround for that?

You can try changing　```#box[$ integral x + y = z $]``` to ```#box(width: 100%)[$ integral x + y = z $]```.

**Comment by Trebor-Huang** [comment]
Is there a consensus on how this should be dealt with? Meanwhile, is there a workaround that does not require explicitly wrapping every block equation in a function call? It seems that when show rules apply, the paragraph break has already been emitted.

**Comment by MrQubo** [comment]
From the user perspective I think it doesn't make sense that lists can attach to the paragraph before and not paragraph after. And from the syntax perspective it would be in my opinion very natural that empty line after the list creates new paragraph and no empty line would not create new paragraph.

That's how blocks of text are already handled. It would also make the behaviour consistent with the behaviour of empty line before the list (i.e. adding empty line before the list adds paragraph separation between the list and preceding paragraph, and not having an empty line before the preceding behaviour will result in a normal line separation).

It seems like lists are handled specially only for implementation reasons and not because of a particular user use-case.

**Comment by jakubkaczor** [comment]
@MrQubo, I thought it was already clear for all participants, that this is how it is expected to work, and that it is only the matter of implementing it. There was only a doubt on whether to allow a figure as a part of a paragraph.

**Comment by aaron-jack-manning** [comment]
> From the user perspective I think it doesn't make sense that lists can attach to the paragraph before and not paragraph after. And from the syntax perspective it would be in my opinion very natural that empty line after the list creates new paragraph and no empty line would not create new paragraph.

I agree, and this is especially true for mathematical equations, and how many existing typesetting systems work. It is very important to be able to clearly express whether or not an equation is part of the paragraph, or at the end of the paragraph, otherwise you end up with equations that are half way through a sentence leading to a new paragraph immediately following it.

**Comment by Matheritasiv** [comment]
One more problem with `#box` is that the box is unbreakable, and for long equations at the end of a page, putting them into a box will cause the whole to move to the next page and leave a big blank space.

**Comment by physy** [comment]
not a pretty way, but if you insert this code into the end of the set and show rules, it works fine.

```typst
#show: it => {
  let cnt = -1
  let processing = 0
  for child in it.children {
    cnt += 1
    if (processing >= 2) {
      processing -= 1
      continue
    } else if (processing == 1) {
      processing -= 1
      context h(-par.first-line-indent.amount) + child
      continue
    }
    child
    if (
      cnt < it.children.len() - 2
        and ((child.func() == math.equation and child.block) or child.func() == list.item or child.func() == enum.item)
        and it.children.at(cnt + 1).func() != parbreak
        and it.children.at(cnt + 2).func() == text
    ) {
      processing = 2
    }
  }
}
```

**Comment by physy** [comment]
```typst
$
  x=1
$
#lorem(10)
$
  y=1
$

#lorem(10)

- A
- B
#lorem(10)

- A
- B

#lorem(10)

+ A
+ B
#lorem(10)
```

gives

<img width="557" alt="Image" src="https://github.com/user-attachments/assets/a5fa5cca-07b7-45f5-b98a-b8f6839ed62e" />

**Comment by iwkjosec** [comment]
#6619 made Typst output tagged PDFs, so simply adjusting the appearance won't fix the incorrect `<P>` tag placement.

**Comment by ericpashman** [comment]
I'm a long-time LaTeX user who recently decided to give Typst a try. I still haven't figured out how to get Typst to produce the behavior described in the OP, in which a new (indented) paragraph of text follows a block of math if and only if I choose to insert a paragraph break between the math block and the following text.

Here's what I've tried:

1. Play with the `first-line-indent` settings. This does not fix the spurious paragraph break, so it just controls whether all text or no text is indented following math blocks.
3. Use the `indenta` package's `fix-indent` function, which claims to do what I want. It produces an error.
4. Use the `#box` hack. Doing this all over the place is a non-starter, but as others have already pointed out, it isn't a viable work-around anyway because it breaks paging.
5. Various other hacks like the one suggest by @physy above. None of these has worked. (The hack above produced hanging indents with the article template I used.)

In my opinion, LaTeX-like behavior is the only sensible choice. Typesetting software that can't typeset a basic article with blocks of math inside a paragraph is useless to me.

**Comment by Enivex** [comment]
> Typesetting software that can't typeset a basic article with blocks of math inside a paragraph is useless to me.

This is unnecessary.

**Comment by Raphael-CV** [comment]
A temporary hack I use, to be inserted manually, not very elegant, but it works for my needs.
Perhaps it may be useful to you.

```typ
#set page("a5", height: auto, columns: 2, margin: 1.5cm)
#set par(justify: true, first-line-indent: (amount: 2em, all: true))

#let no-par(eq: false) = context {
    v(if eq { par.spacing } else { par.leading }, weak: true)
    h(- par.first-line-indent.amount, weak: true)
}

#lorem(10)
- #lorem(3)
- #lorem(3)
#lorem(10)
$ f(t) $
#lorem(10)

#colbreak()

#lorem(10)
- #lorem(3)
- #lorem(3)
#no-par()               // <--
#lorem(10)
$ f(t) $
#no-par(eq: true)       // <--
#lorem(10)
```

<img width="1207" height="650" alt="Image" src="https://github.com/user-attachments/assets/90624ef8-7b52-49b4-a16c-6821c3a293d9" />

**Comment by TheColdPot** [comment]
> not a pretty way, but if you insert this code into the end of the set and show rules, it works fine.
> ```typst
> #show: it => {
>   let cnt = -1
>   let processing = 0
>   for child in it.children {
>     cnt += 1
>     if (processing >= 2) {
>       processing -= 1
>       continue
>     } else if (processing == 1) {
>       processing -= 1
>       context h(-par.first-line-indent.amount) + child
>       continue
>     }
>     child
>     if (
>       cnt < it.children.len() - 2
>         and ((child.func() == math.equation and child.block) or child.func() == list.item or child.func() == enum.item)
>         and it.children.at(cnt + 1).func() != parbreak
>         and it.children.at(cnt + 2).func() == text
>     ) {
>       processing = 2
>     }
>   }
> }
> ```

It still destroys the sematic features by breaking the paragraph before & after. So this does not really solve the problem - we still need a language-level feature for this. Anyhow thanks for the temporal solution <3

At least I think it makes sense to make an equation a part of a paragraph. Simply, I would suggest giving an option to make *display* equations distinct from *block* equations, where the former appears as an independent line but act as an inline box, not a block.

Although being a fan of Typst, I don't know that much of Typst's philosophy. So for community is it a good idea to be some sort of compiler options, bring a breaking change, or introduce a settable parameter of `math.equation`?

**Comment by N3xed** [comment]
Just came across this, the current behavior of equations and lists is pretty bad. I guess manually inserting space is all that's possible at the moment. `first-line-indent: (amount: 2em, all: true)` actually makes the problem worse since this prevents (at least the visual) look that equations are within the same paragraph completely (without other show rule trickery, which I don't think is worth the hassle).

**Comment by reveperdu** [comment]
Block equations might as well be a different case from lists,  I guess this is somewhat specific to chinese typesetting and might be considered part of the language support.

**Comment by YDX-2147483647** [comment]
Here's an improved version of https://github.com/typst/typst/issues/3206#issuecomment-3013274959 that also supports labeled equations. (`Before $ x $ <eq> After` has two `[ ]` after the equation.)

```typst
#set par(first-line-indent: (all: true, amount: 2em))

// This show rule should be put after other show/set rules.
#show: rest => {
  let immediately-after = false

  for it in rest.children {
    // If this text is immediately after special elements (block equations etc.), compensate the first-line indent.
    if immediately-after and it.func() == text {
      context h(-par.first-line-indent.amount)
    }
    it

    // If it is a space, consider we are still immediately after.
    // Notes:
    // - A single `\n` becomes a spaces, and a space between an equation and a label is still a space.
    // - Two consecutive `\n` become a parbreak.
    if it != [ ] {
      immediately-after = false
    }

    // Detect special elements.
    if (
      (it.func() == math.equation and it.block)
        or it.func() == list.item
        or it.func() == enum.item
    ) {
      immediately-after = true
    }
  }
}
```

**Known limitation:** If the text immediately after a special element (a block equation or similar things) starts with `strong`, `link`, or anything that makes `it.func() ≠ text`, then it does not work. Please manually insert a `#h(-2em)` or `#context h(-par.first-line-indent.amount)`.

---

> It still destroys the sematic features by breaking the paragraph before & after

In addition, the vertical spaces around block equations should be [`par.leading`](https://typst.app/docs/reference/model/par/#parameters-leading), but this workaround uses `par.spacing`.

---

> somewhat specific to chinese typesetting

I agree. See https://gap.zhtyp.art/#indent-after-block.

<img width="400" alt="Image" src="https://github.com/user-attachments/assets/417acfd4-1455-4a5d-a467-a469a3ef947a" />

**Comment by Enivex** [comment]
> Block equations might as well be a different case from lists, I guess this is somewhat specific to chinese typesetting and might be considered part of the language support.

I don't understand why chinese typesetting is relevant here? The same problem is present regardless of language.

**Comment by reveperdu** [comment]
> I don't understand why chinese typesetting is relevant here? The same problem is present regardless of language.

For block equations, I notice in some English materials, the paragraph immediately after a block equation isn't indented even if they are (semantically) a new paragraph, which matches the default behavior of `first-line-indent`, without `all:true`, and some English materials don't use first line indents at all.

So while in chinese it's common to distinguish whether the line after the block equations is a new paragraph (if a new paragraph, indent), I thought typesetting in other languages might not enforce such rules, and it won't be a problem if block equations cannot be put into the paragraph, perhaps?

**Comment by Shuenhoy** [comment]
> > I don't understand why chinese typesetting is relevant here? The same problem is present regardless of language.
> 
> For block equations, I notice in some English materials, the paragraph immediately after a block equation isn't indented even if they are (semantically) a new paragraph, which matches the default behavior of `first-line-indent`, without `all:true`, and some English materials don't use first line indents at all.
> 
> So while in chinese it's common to distinguish whether the line after the block equations is a new paragraph (if a new paragraph, indent), I thought typesetting in other languages might not enforce such rules, and it won't be a problem if block equations cannot be put into the paragraph, perhaps?

I don't think it is specific to Chinese typesetting. By semantics, block equations should be part of a paragraph, and I cannot understand why there would be special rule that the next paragraphs of block equations should not be indented.  The most possible cause I guess is that the creators of that materials were unaware.

I randomly open an ACM publication and immediately found they properly distinguish two cases:
<img width="660" height="372" alt="Image" src="https://github.com/user-attachments/assets/7681b7da-b759-4344-99d1-600013097046" />

**Comment by SeSodesa** [comment]
> I'm not really convinced that a figure really makes sense as part of a paragraph.

You are correct, at least from a semantic point of view. In terms of structural PDF or HTML tagging, it makes no sense to have figures inside of paragraph tags.

**Comment by tianyi-smile** [comment]
The `parize` package now gives an experimental solution, offering more flexible control over block-level elements and exploring how this paragraph model affects Typst typesetting.

Performance improvements and bug fixes in version 0.2.1

Example One:

<img width="1133" height="1168" alt="Image" src="https://github.com/user-attachments/assets/ed2034a9-d7ae-4fc4-b031-f9465a18dca1" />

<details>
 <summary>Code:</summary>

  ```typst
#import "@preview/parize:0.2.1": *
#set page(margin: 1cm, height: auto, width: 25cm)
#set par(justify: true)
#table(
  columns: (1fr, 1fr),
  [
    #import "@preview/zebraw:0.6.3": *
    #show: zebraw
    ```typst
    #import "@preview/parize:0.2.1": *
    #set par(first-line-indent: (amount: 2em, all: true))
    #show strong: set text(fill: red, size: 1.2em) // debug
    #set block(stroke: red) // debug
    #show: par-indent.with(
      include-elem: (heading, math.equation, list, enum, terms, ), // Enable paragraph indentation control for heading, math.equation, list, enum and terms
      use-par-leading: true, // Enable paragraph spacing control for list, enum and terms
    )
    #set heading(numbering: "1.")
    = Heading

    *Indented* #lorem(10)

    _Equation Test_. #lorem(8)
    $
      a^2 + b^2 = c^2
    $
    *Unindented* #lorem(10)  // This paragraph remains unindented (no parbreak before equation)

    #lorem(2)

    $
      a^2 + b^2 = c^2
    $

    *Indented* #lorem(10)

    _Tight Lists Test_. #lorem(10)
    + #lorem(2)
    + #lorem(2)
    *Unindented and par-leading* #lorem(8)  // Unindented with paragraph-leading spacing

    = Heading
    *Unindented* #lorem(10)

    + #lorem(2)
    + #lorem(2)

    *Indented* #lorem(10)

    _Non-tight Lists Test_. #lorem(2)

    - #lorem(2)

    - #lorem(2)
    *Unindented* #lorem(2)  // Paragraph following lists without empty line
    ```
  ],
  [
    #show strong: set text(fill: red, size: 1.2em) // debug
    #set block(stroke: red) // debug
    #set par(first-line-indent: (amount: 2em, all: true))
    #show: par-indent.with(
      include-elem: (heading, math.equation, list, enum, terms), // Enable paragraph indentation control for heading, math.equation, list, enum and terms
      use-par-leading: true, // Enable paragraph spacing control for list, enum and terms
    )


    #set heading(numbering: "1.")

    = Heading

    *Indented* #lorem(10)

    _Equation Test_. #lorem(8)
    $
      a^2 + b^2 = c^2
    $
    *Unindented* #lorem(10)  // This paragraph remains unindented (no parbreak before equation)

    #lorem(2)

    $
      a^2 + b^2 = c^2
    $

    *Indented* #lorem(10)

    _Tight Lists Test_. #lorem(10)
    + #lorem(2)
    + #lorem(2)
    *Unindented and par-leading* #lorem(8)  // Unindented with paragraph-leading spacing

    = Heading
    *Unindented* #lorem(10)

    + #lorem(2)
    + #lorem(2)

    *Indented* #lorem(10)

    _Non-tight Lists Test_. #lorem(2)

    - #lorem(2)

    - #lorem(2)
    *Unindented* #lorem(2)  // Paragraph following lists without empty line
  ],
)

  ```
  </details>


Example Two:

<img width="1133" height="702" alt="Image" src="https://github.com/user-attachments/assets/64a57bb7-639d-4b88-8b29-7515610e9e9f" />

<details>
<summary>Code:</summary>

```typst
#import "@preview/parize:0.2.1": *
#set page(margin: 1cm, height: auto, width: 25cm)
#table(
  columns: (1fr, 1fr),
  [
    #import "@preview/zebraw:0.6.3": *
    #show: zebraw
    ```typst
    #import "@preview/parize:0.2.1": *
    #set par(first-line-indent: (amount: 2em, all: true))
    #show strong: set text(fill: red, size: 1.2em) // debug
    #set block(stroke: red) // debug
    #show: par-indent.with(
      include-elem: (), // Enable paragraph indentation control for all block-level elements
      use-par-leading: (
        text-block-leading: (math.equation,),
        block-text-leading: (list, enum, terms, math.equation),
      ),
    )

    _Equation Test_. #lorem(8) *par.leading* // no parbreak after
    $
      a^2 + b^2 = c^2
    $
    *Unindented + par.leading* #lorem(10) // Paragraph following equation without empty line

    _Equation Test_. #lorem(8) *par.spacing*

    $
      a^2 + b^2 = c^2
    $

    *Indented + par.spacing* #lorem(10)
    ```
  ],
  [
    #set par(first-line-indent: (amount: 2em, all: true))
    #show strong: set text(fill: red, size: 1.2em) // debug
    #set block(stroke: red) // debug
    #show: par-indent.with(
      use-par-leading: (
        text-block-leading: (math.equation,),
        block-text-leading: (list, enum, terms, math.equation),
      ),
    )
    _Equation Test_. #lorem(8) *par.leading* // no parbreak after
    $
      a^2 + b^2 = c^2
    $
    *Unindented + par.leading* #lorem(10) // Paragraph following equation without empty line


    _Equation Test_. #lorem(8) *par.spacing*

    $
      a^2 + b^2 = c^2
    $

    *Indented + par.spacing* #lorem(10)
  ],
)
```

</details>


See https://typst.app/universe/package/parize for more configurations.

See [ex-in-elems-test.typ](https://github.com/tianyi-smile/parize/blob/main/tests/ex-in-elems-test.typ) for details on using `par-indent` with the `rect` element when Paragraph Indentation or Paragraph Spacing is enabled.

See also [inlinable-block.pdf](https://github.com/tianyi-smile/parize/blob/main/examples/inlinable-block.pdf) ([source](https://github.com/tianyi-smile/parize/blob/main/examples/inlinable-block.typ)) for details on using `par-indent` with custom block-level containers that can be treated as part of a paragraph.

**Comment by wade-cheng** [comment]
> > I'm not really convinced that a figure really makes sense as part of a paragraph.
> 
> You are correct, at least from a semantic point of view. In terms of structural PDF or HTML tagging, it makes no sense to have figures inside of paragraph tags.

I'm not sure what definition of semantics are being used here, but I agree with the given opinions that figures and math can be part of a paragraph. It's like how you can have a sentence that starts out as regular text and ends as a block level equation---that's still seen as a sentence, in mathy works. In the same vein, a paragraph is "just," "semantically," a bunch of stuff with the same idea (is how I see it). So, figures and math can definitely be part of an idea, just as how they can be part of a sentence.

**Comment by SeSodesa** [comment]
> > > I'm not really convinced that a figure really makes sense as part of a paragraph.
> > 
> > 
> > You are correct, at least from a semantic point of view. In terms of structural PDF or HTML tagging, it makes no sense to have figures inside of paragraph tags.
> 
> I'm not sure what definition of semantics are being used here, but I agree with the given opinions that figures and math can be part of a paragraph. It's like how you can have a sentence that starts out as regular text and ends as a block level equation---that's still seen as a sentence, in mathy works. In the same vein, a paragraph is "just," "semantically," a bunch of stuff with the same idea (is how I see it). So, figures and math can definitely be part of an idea, just as how they can be part of a sentence.

I agree on the math part, since an equation can be a part of a sentence and a paragraph is a collection of sentences in natural language. But I would not include figures into sentences. References to figures do make sense as a part of a sentence.

**Comment by AstraArcane** [comment]
Having lists and display equation within a single paragraph is extremely common in academic papers and lecture notes. The fact that Typst doesn't natively support this is actually unusual‌, so let's stop debating whether this feature is necessary and focus on how to implement it.

Currently, a display math block forces the subsequent text into a new paragraph. This happens because Typst treats math blocks and paragraphs differently; inserting a math block essentially splits what should be a single, continuous paragraph. Writing a separate package just to patch the indentation issue feels like a hack. As a typesetting tool, Typst should address this at the language level for better usability—just like LaTeX, where a blank line creates a new paragraph and no blank line keeps the text in the same one. The block itself shouldn't implicitly insert a `#parbreak`; paragraph breaking should be handled during compilation.
